File: Core/PROGETTIHWSW/ReleUSB.cs
using System;
using System.Collections.Generic;
using System.IO.Ports;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Core.PROGETTIHWSW
{
    public class ReleUSB
    {
        private string COM;
        private SerialPort SERIAL;

        public ReleUSB(string comPort, SerialPort serialPort)
        {
            COM = comPort;
            SERIAL = serialPort;
        }

        public void Connect()
        {
            try
            {
                SERIAL.PortName = COM;
                SERIAL.BaudRate = 115200;
                SERIAL.Parity = Parity.None;
                SERIAL.DataBits = 8;
                SERIAL.StopBits = StopBits.One;
                SERIAL.ReadTimeout = 500;
                SERIAL.WriteTimeout = 500;
                SERIAL.Handshake = Handshake.None;
                try
                {
                    SERIAL.Open();
                }
                catch (Exception ex)
                {
                    Console.WriteLine("ReleUSB.Connect : " + ex.Message);
                    throw ex;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("ReleUSB.Connect: " + ex.Message);
                Console.WriteLine("ReleUSB.Connect: " + ex.StackTrace);
                throw ex;
            }
        }

        public void InviaComando(string command)
        {
            try
            {
                SERIAL.Write(command);
            }
            catch (Exception ex)
            {
                Console.WriteLine("ReleUSB.InviaComando:" + ex.Message);
                throw ex;
            }
        }

        public string RiceviDati()
        {
            string ret = "";
            try
            {
                ret = SERIAL.ReadExisting();
            }
            catch (Exception ex)
            {
                Console.WriteLine("ReleUSB.RiceviDati:" + ex.Message);
                throw ex;
            }
            return ret;
        }

        private void Clean()
        {
            try
            {
                //-----Cleaning the Buffer
                SERIAL.DiscardOutBuffer();
                SERIAL.DiscardInBuffer();
                //-----Finish Cleaning
            }
            catch (Exception ex)
            {
                Console.WriteLine("LedRGBCOM.Clean: " + ex.Message);
                Console.WriteLine("LedRGBCOM.Clean: " + ex.StackTrace);
                throw ex;
            }
        }

        public List<string> GetCmdAndResponse(int sensorNumber)
        {
            List<string> lista = new List<string>();
            try
            {
                if (sensorNumber == 1)
                {
                    lista.Add(Comandi.AccendiRelay1);
                    lista.Add(Risposte.AccendiRelay1);
                    lista.Add(Comandi.SpegniRelay1);
                    lista.Add(Risposte.SpegniRelay1);
                    lista.Add(Comandi.Ingresso1);
                    lista.Add(Risposte.Ingresso1Alto);
                    lista.Add(Risposte.Ingresso1Basso);
                }
                else if (sensorNumber == 2)
                {
                    lista.Add(Comandi.AccendiRelay2);
                    lista.Add(Risposte.AccendiRelay2);
                    lista.Add(Comandi.SpegniRelay2);
                    lista.Add(Risposte.SpegniRelay2);
                    lista.Add(Comandi.Ingresso2);
                    lista.Add(Risposte.Ingresso2Alto);
                    lista.Add(Risposte.Ingresso2Basso);
                }
                else if (sensorNumber == 3)
                {
                    lista.Add(Comandi.AccendiRelay3);
                    lista.Add(Risposte.AccendiRelay3);
                    lista.Add(Comandi.SpegniRelay3);
                    lista.Add(Risposte.SpegniRelay3);
                    lista.Add(Comandi.Ingresso3);
                    lista.Add(Risposte.Ingresso3Alto);
                    lista.Add(Risposte.Ingresso3Basso);
                }
                else if (sensorNumber == 4)
                {
                    lista.Add(Comandi.AccendiRelay4);
                    lista.Add(Risposte.AccendiRelay4);
                    lista.Add(Comandi.SpegniRelay4);
                    lista.Add(Risposte.SpegniRelay4);
                    lista.Add(Comandi.Ingresso4);
                    lista.Add(Risposte.Ingresso4Alto);
                    lista.Add(Risposte.Ingresso4Basso);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("ReleUSB.GetCmdAndResponse:"+ex.Message);
                throw ex;
            }

            return lista;
        }
    }
}

File: ReactionTrainingPY/it/napalm/core/progettihwsw/ReleUSB.py
'''
Created on 18/mar/2015

@author: Luigi
'''
import serial
from it.napalm.core.object.Event import Event
from it.napalm.core.progettihwsw.Comandi import Comandi
from it.napalm.core.progettihwsw.Risposte import Risposte
class ReleUSB(object):
    COM = ''
    SERIAL = 0
    TEST = '1'
    
    def __init__(self, comPort):
        #print('ISt')
        self.COM = comPort
        self.EventoDatoRicevuto = Event()
        
    def Connect(self):
        try:
            self.SERIAL = serial.Serial(
                port=self.COM,
                baudrate=115200,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE,
                bytesize=serial.EIGHTBITS
            )
            if self.SERIAL.isOpen() :
                self.SERIAL.close()
            self.SERIAL.open()
            #print(self.TEST)
            t = LetturaSeriale(self)
            t.EventoDatoRicevuto += DatoRicevutoRele
            t.start()
            #time.sleep(1)
            #self.InviaComando('6')
            
            #while True:
            #    time.sleep(.5)
                #print(self.TEST)
        except Exception as e :
            print("errore serial: " +  str(e))
            
            
    def Clean(self):
        try:
            self.SERIAL.flushInput()
            self.SERIAL.flushOutput()
        except Exception as e :
            print("errore Clean: " +  str(e))
            
    def InviaComando(self, command):
        try:
            print('invio:' + command)
            self.Clean()
            self.SERIAL.write(command.encode('utf-8'))
        except Exception as e :
            print("errore InviaComando: " +  str(e))
            
    def Close(self):
        try:
            if self.SERIAL.isOpen() :
                self.SERIAL.close()
        except Exception as e :
            print("errore Close: " +  str(e))
            
    def GetCmdAndResponse(self, sensorNumber):
        lista = []
        try:
            print(sensorNumber)
            if sensorNumber == 1:
                lista.append(Comandi.AccendiRelay1);
                lista.append(Risposte.AccendiRelay1);
                lista.append(Comandi.SpegniRelay1);
                lista.append(Risposte.SpegniRelay1);
                lista.append(Comandi.Ingresso1);
                lista.append(Risposte.Ingresso1Alto);
                lista.append(Risposte.Ingresso1Basso);
            elif sensorNumber == 2:
                lista.append(Comandi.AccendiRelay2);
                lista.append(Risposte.AccendiRelay2);
                lista.append(Comandi.SpegniRelay2);
                lista.append(Risposte.SpegniRelay2);
                lista.append(Comandi.Ingresso2);
                lista.append(Risposte.Ingresso2Alto);
                lista.append(Risposte.Ingresso2Basso);
            elif sensorNumber == 3:
                lista[Comandi.AccendiRelay3];
                lista[Risposte.AccendiRelay3];
                lista[Comandi.SpegniRelay3];
                lista[Risposte.SpegniRelay3];
                lista[Comandi.Ingresso3];
                lista[Risposte.Ingresso3Alto];
                lista[Risposte.Ingresso3Basso];
            elif sensorNumber == 4:
                lista[Comandi.AccendiRelay4];
                lista[Risposte.AccendiRelay4];
                lista[Comandi.SpegniRelay4];
                lista[Risposte.SpegniRelay4];
                lista[Comandi.Ingresso4];
                lista[Risposte.Ingresso4Alto];
                lista[Risposte.Ingresso4Basso];
        except Exception as e :
            print("errore GetCmdAndResponse: " +  str(e))
            
        return lista
    
    
def DatoRicevutoRele(dato, rele):
    rele.EventoDatoRicevuto(dato)     
        
import threading 
class LetturaSeriale(threading.Thread):
    RELE = 0;
    def __init__(self, releUSB):
        threading.Thread.__init__(self)
        self.RELE = releUSB
        self.EventoDatoRicevuto = Event()
    
    def run(self):
        try:
            while True:
                last = ''
                for byte in self.RELE.SERIAL.read(self.RELE.SERIAL.inWaiting()): 
                    last += chr(byte)
                    test = bytes(last, "utf-8").decode("utf-8")
                    self.EventoDatoRicevuto(test, self.RELE)
                    if len(last) > 0:
                        last = ''
        except Exception as e :
            print("errore run LetturaSeriale:  " +  str(e))


def RicevoDato(dato):
    if dato.isalpha() :
        print ("Ricevo: " + dato)

if __name__ == "__main__":
    rele = ReleUSB('COM14')
    rele.EventoDatoRicevuto += RicevoDato

    rele.Connect()

